### ISSUE TITLE
[Datasets] Support for ingest for distributed training

### ISSUE THREAD

**Issue by amogkam** [ray-project/ray#18287]
### Describe your feature request
I want to use Ray Datasets for parallel training. This requires me to do the following:

1. Create N training workers (Ray actors)
2. Split the dataset into N equal size shards. The Dataset should be padded so that the number of rows is divisible by N. Ideally, locality hints would be supported here to minimize the cross-node data transfer.
3. Send each shard to the corresponding training worker
4. Each training worker iterates through the dataset `num_epochs` times, 
    4b. The training workers know when an epoch has ended.
5. After each epoch, the dataset is shuffled across all workers.

2 is necessary for synchronous training like PyTorch DDP, otherwise workers will hang if there is a mismatch in shard size. 4b is necessary to know when to place per-epoch operations (such as logging metrics to MLflow for example).

Currently, 2 and 4b are not supported by Ray Datasets out of the box.

I've spent quite a bit of time trying to get this workflow to work via combinations of `repeat`, `split_at_indices`, `union`, `random_shuffle`, but I can't get this to work.

Can this workflow either be supported out of the box, or an example be provided showing how to do this and use Ray Datasets with PyTorch for example?

**Comment by clarkzinzow** [comment]
Sounds great @amogkam! This is the API that I have had in mind for a while.

FYI, unless I'm misunderstanding the data model, you should be able to satisfy 4(b) right now via `DatasetPipeline.iter_datasets()`:
```python
@ray.remote
class TrainingWorker:
    def train(shard_pipeline):
        model = Model()
        for epoch, shard_dataset in enumerate(shard_pipeline.iter_datasets()):
            for batch in shard_dataset.iter_batches():
                model(batch)
            print("Epoch: ", epoch)
```

**Comment by amogkam** [comment]
Ah I didn't even realize that you can use `iter_datasets` with `repeat` like that. Yes that should satisfy 4b, though from a usability perspective I would prefer to define the number of epochs inside the training function itself, rather than outside.  It seems like `iter_datasets` is a developer API though right?

**Comment by clarkzinzow** [comment]
(2) is satisfied via equal splitting with minimal (`num_rows % num_shards` rows dropped) truncation (which I believe met the underlying requirements here without adding support for sample padding), and (4[b]) is supported via `.iter_epochs()`, so I'm closing this for now.
